### ISSUE TITLE
Load conf via ENV variable for docker containers

### ISSUE THREAD

**Issue by bwaz** [coredns/coredns#514]
Not sure if this could be an enhancement or if it's required?

To load a custom config we need to bind the volume to the host 
```
rwasniak/work/src/github.com/miekg/coredns$ docker run -v "/Conf:/Conf" bob/coredns -conf /Conf/C
```
We could have coredns take the input from a named env variable, just an idea.

**Comment by miekg** [comment]
You can pipe the config to CoreDNS (search for: caddy pipe config file [IIRC]). This however means that transparant restarts don't work.

**Comment by bwaz** [comment]
Yes, here is an example
```

rwasniak/work/src/github.com/miekg/coredns$ echo "$COREFILE" | docker run -i  bob/coredns 
.:53
2017/02/09 19:59:03 [INFO] CoreDNS-004
CoreDNS-004
^C2017/02/09 19:59:04 [INFO] SIGINT: Shutting down
rwasniak/work/src/github.com/miekg/coredns$ 
rwasniak/work/src/github.com/miekg/coredns$ echo "$COREFILE" | docker run -i  bob/coredns -conf stdin
.:54
2017/02/09 19:58:58 [INFO] CoreDNS-004
CoreDNS-004

```

**Comment by miekg** [comment]
I think I actually brought this up on the Caddy github from @mholt, but he was against it. We can of course add something that would make that easier. I.e. I'm not against making runing in Docker harder than it should be.

**Comment by mholt** [comment]
Which issue was this exactly? I don't know Docker and I'm having a hard time making the link here... 🤔

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Load conf via E..." ]
>Which issue was this exactly? I don't know Docker and I'm having a hard time making the link here... 🤔

reading the entire Caddy/Core-file from a env var. This helps in docker because
you don't have to store anything inside the container, making the docker
container more versatile.

**Comment by bwaz** [comment]
It's minimal effort to have a bash script to be a wrapper for core dns, the presence of an env variable could use the pipe feature.

**Comment by bwaz** [comment]
PR will be soon:
```
rwasniak/work/src/github.com/miekg/coredns$ echo $COREFILE
.:55 { log stdout }
rwasniak/work/src/github.com/miekg/coredns$ docker run -it -e "zCOREFILE=$COREFILE" $USER/coredns
+ set -- /coredns
+ [ -z  ]
+ exec /coredns
.:53
2017/02/10 15:43:53 [INFO] CoreDNS-005
CoreDNS-005
^C2017/02/10 15:44:02 [INFO] SIGINT: Shutting down
rwasniak/work/src/github.com/miekg/coredns$ docker run -it -e "COREFILE=$COREFILE" $USER/coredns
+ set -- /coredns
+ [ -z .:55 {
log stdout
}
 ]
+ /coredns -conf stdin
+ exec echo .:55 {
log stdout
}

.:55
2017/02/10 15:44:07 [INFO] CoreDNS-005
CoreDNS-005
^C2017/02/10 15:44:09 [INFO] SIGINT: Shutting down


```

**Comment by mholt** [comment]
I suppose a Caddyfile loader plugin could be written to do it, but I dunno if it would come standard with Caddy. It would simply look for a certain env var and use that as its Caddyfile (or Corefile). But it would be suuuuper easy to write.

**Comment by bwaz** [comment]
https://github.com/miekg/coredns/pull/518

**Comment by bwaz** [comment]
@mholt  It would be a nice addition to look for env variable,  #518  is a stop-gap solution.

**Comment by mholt** [comment]
@bwaz Here's the instructions for writing a Caddyfile loader plugin: https://github.com/mholt/caddy/wiki/Writing-a-Plugin:-Caddyfile-Loader

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Load conf via E..." ]
>@mholt  It would be a nice addition to look for env variable,  #518  is a stop-gap solution.

Haven't looked into writing/hooking up a custom loader - that sounds most
elegant though.

**Comment by bwaz** [comment]
Do we or do we not use #518? Env reading could be added to Caddy, but this is working for now and we can defer the feature for later.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [miekg/coredns] Load conf via E..." ]
>Do we or do we not use #518? Env reading could be added to Caddy, but this is working for now and we can defer the feature for later.

I'm inclining to 'not using #518'. Doing it in Go is the better approach, doing
it in a custom loader is even better - but that might be more work. I'm not sure
if we can do it without using a custom loader though...

**Comment by bwaz** [comment]
Closed #518 PR, for reference: https://github.com/miekg/coredns/pull/518/files

**Comment by miekg** [comment]
issue seems stale
